src/java.base/share/classes/jdk/internal/foreign/abi/fallback/LibFallback.java

```diff
@@ -90,10 +90,11 @@ private static boolean tryLoadLibrary() {
      * @see jdk.internal.foreign.abi.CapturableState
      */
     static void doDowncall(MemorySegment cif, MemorySegment target, MemorySegment retPtr, MemorySegment argPtrs,
-                           MemorySegment capturedState, int capturedStateMask,
+                           Object captureStateHeapBase, MemorySegment capturedState, int capturedStateMask,
```

Comment:
I find it a bit weird that this method is receiving both the captured segment base and the segment itself. It almost seems as if the checks we do in `FallbackLinker` should be done here?

Reply:
I see what you mean, and I agree it looks a bit strange. but we also pass the array of heap bases separately here (which is more or less a forced move due to how the libffi API works). So, I think the current code is more inline with that. The `LibFallback` class is only supposed to be a thin wrapper around the native methods in `fallbackLinker.c`. The actual logic is all in `FallbackLinker`.